sdk_refactor_design_proposal.md

```diff
@@ -0,0 +1,392 @@
+
+This functionality will be implemented using the [Factory pattern](https://refactoring.guru/design-patterns/factory-method). Each factory-generated class will have its own implementation and validation mechanism. This will make the implementation easy to maintain as this makes it easy to swap algorithms used inside an object during runtime. That way, we can isolate the implementation details from the *Client*. If we choose to implement a new strategy, for example fetching values from environment values, this will not require a change in the Client.
+
+##### Proposed flow
+
+```python
+config = client.Configuration(action=file)
+creds = client.Credental(action=keystore)
+client. conjur.ConjurClient(config, creds, debug, ssl_verify)
+```
+
+```python
+# data object representing the configuration data 
+# needed to connect to the Conjur server
+class ConfigurationModel(action = None,
+                         url = None,
+                         account = None,
+                         ca_bundle = None,
+                         ssl_verify = True):
+
+# configuration factory deciding which flow to run based on user input
+class Configuration(action=None, 
+                    ssl_verify=None)
+
+	if configuration.action == "file":
+		return ConfigurationFromFile(configuration)
+  elif configuration.action == "env":
+		return ConfigurationFromEnv(configuration)
+  
+class ConfigurationFromFile(ConfigurationStrategy)
+	def __init__(url=None, 
+               account=None,
+               ca_bundle=None,
+               ssl_verify=True)
+		
+    self.validate()
+    
+  # validates all parameters were given
+  def validate():
+    if not self.url:
+      raise ValueError("Error: Appliance URL not provided") 
+    # ...
+```
+
+Note: the same pattern will be implemented for Credentials. 
+
+For both Configuration and Credential flows, we will still support the ability for users to build these objects themselves. If so, we won’t fetch the values from disk or any other supported mediums for them. For example:
+
+```python
+config = client.Configuration(url="https://someserver", account="cucumber", cert_file="/some/certfile")
+creds = client.Credental(machine="https://someserver", username="someuser", api_key="123")
+client. conjur.ConjurClient(config, creds, debug, ssl_verify)
+```
+
+We will *not* create the conjurrc and the netrc for the user as is true in the current state of the SDK. 
+
+We will *not* support partial states as is true in the current SDK. For example, we will not support the flow if a user provides `client.Configuration(action=file, url="https://someserver", account="", cert_file="/some/certfile")`. In this flow we will fail with relevant error.
+
+
+
+##### API and commands
+
+The following code excerpt was taken from the current implementation of *Client* and *API* classes:
+
+```python
+class Client()
```

Comment:
The name `client` is very confusing to me... client of what?

Reply:
I agree. You are referencing the current SDK logic. 

I wanted to update this to `ConjurClient`. WDYT?